pkg/svc/zserver.go

```diff
@@ -2,51 +2,196 @@ package svc
-// ~~~~~~~~~~~~~~~~~~~~~~~~~ A BRIEF DEVELOPMENT GUIDE ~~~~~~~~~~~~~~~~~~~~~~~~~
-//
-// TODO: Extend the MyApp service by defining new RPCs and
-// and message types in the pb/service.proto file. These RPCs and messages
-// compose the API for your service. After modifying the proto schema in
-// pb/service.proto, call "make protobuf" to regenerate the protobuf files.
-//
-// TODO: Create an implementation of the MyApp server
-// interface. This interface is generated by the protobuf compiler and exists
-// inside the pb/service.pb.go file. The "server" struct already provides an
-// implementation of MyApp server interface service, but only
-// for the GetVersion function. You will need to implement any new RPCs you
-// add to your protobuf schema.
-//
-// TODO: Update the GetVersion function when newer versions of your service
-// become available. Feel free to change GetVersion to better-suit how your
-// versioning system, or get rid of it entirely. GetVersion helps make up a
-// simple "starter" example that allows an end-to-end example. It is not
-// required.
-//
-// TODO: Oh yeah, delete this guide when you no longer need it.
-//
-// ~~~~~~~~~~~~~~~~~~~~~~~~~~~ FAREWELL AND GOOD LUCK ~~~~~~~~~~~~~~~~~~~~~~~~~~
-// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-
 const (
 	// version is the current version of the service
 	version = "0.0.1"
 )
 
 // Default implementation of the MyApp server interface
-type server struct{}
+type server struct {
+	pb.UnimplementedMyAppServer
+	Description string
+	Timestamp   time.Time
+	Requests    int64
+	mtx         sync.RWMutex
+}
+
+func GRPCConnect() (*grpc.ClientConn, error) {
+	var opts []grpc.DialOption
+	opts = append(opts, grpc.WithInsecure())
+	addr := viper.GetString("addr")
+	conn, err := grpc.Dial(addr, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return conn, nil
+}
 
 // GetVersion returns the current version of the service
-func (server) GetVersion(context.Context, *empty.Empty) (*pb.VersionResponse, error) {
+func (s *server) GetVersion(context.Context, *empty.Empty) (*pb.VersionResponse, error) {
 	return &pb.VersionResponse{Version: version}, nil
+	//TODO use hash commit as version
+}
+
+func (s *server) UpdateDescription(ctx context.Context, req *pb.UpdateDescriptionRequest) (*pb.UpdateDescriptionResponse, error) {
+	if req.GetDescription() == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "Description can't be empty")
+	}
+	if req.GetService() != 1 {
+		conn, err := GRPCConnect()
+		if err != nil {
+			return nil, err
+		}
+		defer conn.Close()
+		client := rpb.NewMyResponderClient(conn)
+		resp, err := client.UpdateDescription(ctx, &rpb.UpdateDescriptionRequest{Description: req.Description, Service: req.Service})
+		if err != nil {
+			return nil, err
+		}
+		return &pb.UpdateDescriptionResponse{Description: resp.Description}, nil
+	}
+	s.IncRequests()
+	s.Description = req.Description
```

Comment:
What's about concurrency? This field is using in function `GetDescription` as well. Moreover, nobody can deny to call `UpdateDescription` as many times as it possible, in parallel

Reply:
Sorry, forget about concurrency in that case.